src/transaction.rs

```diff
@@ -307,13 +315,88 @@ impl Transaction {
 
     pub fn set_receipts_root(&mut self, root: Bytes32) -> Option<Bytes32> {
         match self {
-            Self::Script { receipts_root, .. } => Some(std::mem::replace(receipts_root, root)),
+            Self::Script { receipts_root, .. } => Some(mem::replace(receipts_root, root)),
 
             _ => None,
         }
     }
 
+    /// Append a new unsigned input to the transaction.
+    ///
+    /// When the transaction is constructed, [`Transaction::sign_inputs`] should
+    /// be called for every secret key used with this method.
+    ///
+    /// The production of the signatures can be done only after the full
+    /// transaction skeleton is built because the input of the hash message
+    /// is the ID of the final transaction.
+    #[cfg(feature = "std")]
+    pub fn add_unsigned_coin_input(
+        &mut self,
+        utxo_id: UtxoId,
+        owner: &PublicKey,
+        amount: Word,
+        color: Color,
+        maturity: Word,
+        predicate: Vec<u8>,
+        predicate_data: Vec<u8>,
+    ) {
+        let owner = Input::coin_owner(owner);
+
+        let witness_index = self.witnesses().len() as u8;
+        let input = Input::coin(
+            utxo_id,
+            owner,
+            amount,
+            color,
+            witness_index,
+            maturity,
+            predicate,
+            predicate_data,
+        );
+
+        self._add_witness(Witness::default());
+        self._add_input(input);
+    }
+
+    /// For all inputs of type `coin`, check if its `owner` equals the public
+    /// counterpart of the provided key. Sign all matches.
+    #[cfg(feature = "std")]
+    pub fn sign_inputs(&mut self, secret: &SecretKey) {
+        let pk = PublicKey::from(secret);
+        let pk = Input::coin_owner(&pk);
+        let id = self.id();
+
+        // Safety: checked length
+        let message = unsafe { Message::as_ref_unchecked(id.as_ref()) };
+
+        let signature = Signature::sign(secret, message);
+
+        let (inputs, witnesses) = match self {
+            Self::Script {
+                inputs, witnesses, ..
+            } => (inputs, witnesses),
+            Self::Create {
+                inputs, witnesses, ..
+            } => (inputs, witnesses),
+        };
+
+        inputs
+            .iter()
+            .filter_map(|input| match input {
+                Input::Coin {
+                    owner,
+                    witness_index,
+                    ..
+                } if owner == &pk => Some(*witness_index as usize),
+                _ => None,
+            })
```

Comment:
```suggestion
            })
            .dedupe()
```

Reply:
*fixed in https://github.com/FuelLabs/fuel-tx/pull/57/commits/bc11f3c307175412c752b7ede457213521e77ae0